url-plugins/src/main/java/co/cask/hydrator/plugin/realtime/UrlFetchRealtimeSource.java

```diff
@@ -0,0 +1,130 @@
+import co.cask.cdap.api.data.schema.Schema;
+import co.cask.cdap.etl.api.Emitter;
+import co.cask.cdap.etl.api.realtime.RealtimeSource;
+import co.cask.cdap.etl.api.realtime.SourceState;
+import co.cask.hydrator.plugin.realtime.config.UrlFetchRealtimeSourceConfig;
+import com.google.common.base.Joiner;
+import com.google.common.base.Strings;
+import com.google.common.io.CharStreams;
+
+import java.io.InputStreamReader;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nullable;
+
+/**
+ * Real Time Source to fetch data from a url.
+ */
+@Plugin(type = RealtimeSource.PLUGIN_TYPE)
+@Name("URLFetch")
+@Description("Fetch data from an external URL at a regular interval.")
+public class UrlFetchRealtimeSource extends RealtimeSource<StructuredRecord> {
+  private static final int TIMEOUT = 60 * 1000;
+  private static final String METHOD = "GET";
+  private static final String POLL_TIME_STATE_KEY = "lastPollTime";
+  private static final Schema SCHEMA = Schema.recordOf(
+    "event",
+    Schema.Field.of("ts", Schema.of(Schema.Type.LONG)),
+    Schema.Field.of("url", Schema.of(Schema.Type.STRING)),
+    Schema.Field.of("responseCode", Schema.of(Schema.Type.INT)),
+    Schema.Field.of("headers", Schema.mapOf(Schema.of(Schema.Type.STRING), Schema.of(Schema.Type.STRING))),
+    Schema.Field.of("body", Schema.of(Schema.Type.STRING))
+  );
+
+  private final UrlFetchRealtimeSourceConfig config;
+
+  public UrlFetchRealtimeSource(UrlFetchRealtimeSourceConfig config) {
+    this.config = config;
+  }
+
+  @Nullable
+  @Override
+  public SourceState poll(Emitter<StructuredRecord> writer, SourceState currentState) throws Exception {
+    byte[] lastPollTimeBytes = currentState.getState(POLL_TIME_STATE_KEY);
+    if (lastPollTimeBytes != null) {
+      long lastPollTime = Bytes.toLong(lastPollTimeBytes);
+      long currentPollTime = System.currentTimeMillis() / 1000;
+      long diff = currentPollTime - lastPollTime;
+      if (config.getIntervalInSeconds() - diff > 0) {
+        TimeUnit.SECONDS.sleep(1L);
+        return currentState;
+      }
+    }
+    try {
+      URL url = new URL(config.getUrl());
+      String response = "";
+      HttpURLConnection connection = (HttpURLConnection) url.openConnection();
+      connection.setRequestMethod(METHOD);
+      connection.setConnectTimeout(TIMEOUT);
+      connection.setInstanceFollowRedirects(config.shouldFollowRedirects());
+      if (config.hasCustomRequestHeaders()) {
+        // Set additional request headers if needed
+        for (Map.Entry<String, String> requestHeader : config.getRequestHeadersMap().entrySet()) {
+          connection.setRequestProperty(requestHeader.getKey(), requestHeader.getValue());
+        }
+      }
+      try {
+        if (connection.getResponseCode() >= 400 && connection.getErrorStream() != null) {
```

Comment:
is error stream only there if there is a 400+ response code?  If not, it would be better just to use it if its not null instead of checking for response code.